scheduler/src/tbf/ratelimit.clj

```diff
@@ -0,0 +1,101 @@
+;;
+;; Copyright (c) Two Sigma Open Source, LLC
+;;
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;  http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns tbf.ratelimit
+  (:require
+    [tbf.core :as tbf-ns])
+  (:import (com.google.common.cache CacheBuilder Cache)
+           (java.util.concurrent TimeUnit)))
+
+(defprotocol RateLimiter
+  (earn-tokens [this key]
+    "Account for any earned tokens for the requested key.")
+  (in-debt [this key]
+    "Boolean indicating if the rate limit is currently in debt. Note that this is read-only and does
+    not earn tokens. For an up-to-date value, you will want to earn tokens first.")
+  (take-time-until [this key tokens]
+    "Request the number of tokens. Returns the number of milliseconds one should wait until those tokens should be spent.
+    (I.e. if currently in debt at the time of the request, before the request is actually processed, this will tell
+    you how long until you get out of debt.)")
+  (time-until [this key]
+    "Make a rate limiter module. Each bucket has a maximum capacity, earns at the requested rate per minute.
+     Old unused buckets expire after the given number of minutes."))
+
+(defn get-tbf
+      "Get or create a TBF for the requested key. Buckets start off half-full, so users whose first request is a lot of
+      fast requests don't get throttled immediately. The fact that we don't have a TBF says that we are either just starting
+      up (not likely) or haven't gotten anything from them in hours."
+      [limiter key]
+      (.get (:cache limiter) key (proxy [Callable] []
+                                   (call []
+                                     (tbf-ns/make (/ (:rate-per-minute limiter) 60000.)
+                                                  (:max-tokens limiter)
+                                                  (System/currentTimeMillis)
```

Comment:
Please use `clj-time.core/now`

Reply:
From clj-time.core:
"" If you are using Java 8 or later, consider using the built-in Java Time instead of Joda Time -- and look at clojure.java-time if you want a Clojure wrapper for that.""

Whats the benefit of going thorugh clojure-java-time or clj-time.core in this use case --- a millisecond counter --- where I'm not doing any date-time logic here: no dates, no hours-minutes-seconds, no timezones, no year-month-day.

Reply:
You can easily mock the call to `now` in your unit tests.

Reply:
We might need a tiebreaker reviewer :)

Reply:
I will also look at the test code for `TBF-RateLimiter` and let you know if I see anything odd.